feat: custom byline interface

allysonsouza opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

This Pull Request implements the feature to allow the selection of author tokens from authors selected on Co-Authors Plus to compound a custom byline with them.

### How to test the changes in this Pull Request:

1. Go to a post edit page and assign authors to the post on co-authors plus panel
2. Below it, you will see the Newspack Custom Byline panel. Set the toggle to active to enable custom byline.
3. You will see the custom byline field, which is a `div` with `contenteditable`. Below it you should see the authors assigned to the post as buttons that will insert the tokens on the content when clicked.
4. Click on buttons to add the token to the content
5. Click on buttons inside the byline content to remove token
6. Also, the tokens should be removed when editions are made to then with delete/backspace for example
7. Save the post, reload the page and see if the custom byline saved is what is expected
8. Check if the custom byline is editable as expected too after saving it

**Things to check:**

- The authors set on Co-Authors Plus are being displayed as tokens to be used on Custom Byline?
- When I click on token "+", the author token is inserted at the end of the byline?
- Does the token get removed after a change on it's content is made (by using delete, backspace or by adding additional characters)?
- Does the token get removed when I click on "x" on it after inserted into the byline?
- Does additional space after the token is inserted to allow more text to be typed after it?
- Does the byline get save properly?
- After saving, does the available tokens (the ones that aren't in use) are correctly displayed below the byline field?

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

allysonsouza committed:
feat: modify textarea to RichText and add TagInlineBlock with dummy authors

allysonsouza committed:
feat: fetch authors from co-authors plus and set tokens on BylinesSettingsPanel

allysonsouza committed:
chore: improve inline documentation

allysonsouza committed:
feat: [WIP] add author tokens on click

allysonsouza committed:
feat: append users to the end of the byline

allysonsouza committed:
feat: add author tokens to custom byline

allysonsouza committed:
feat: handling tokens with vanilla JS (WIP)

allysonsouza committed:
feat: useRef to manage tokens in use

allysonsouza committed:
chore: improve inline docs

allysonsouza committed:
feat: check custom byline after DOM ready

allysonsouza committed:
chore: addEventListener trough ref instead of DOM directly

Also adds the return of a function on useEffect to allow React cleanup on unmount

allysonsouza committed:
feat: display icons on tokens (plus to insert and close to remove)

allysonsouza committed:
feat: add mutation observer to remove tokens on erasing with backspace/delete

allysonsouza committed:
chore: enhance inline documentation

claudiulodro reviewed (dismissed):
Overall I think this is coming together! The UX of it feels good. I noticed a few quirks in my testing:

1. In general it's kind of hard to read once there are a few authors in there:

<img width="303" alt="Screenshot 2025-02-13 at 1 29 49 PM" src="https://github.com/user-attachments/assets/d2d8b522-4afa-4ac2-9e85-401436d4595b" />

Not sure if we should reduce the size of the author boxes or how best to handle that. What do you think, @thomasguillot?

2. Using the arrow keys, you can go into one of the boxes and then type to modify the content. I think it should not be editable.

<img width="294" alt="Screenshot 2025-02-13 at 1 31 01 PM" src="https://github.com/user-attachments/assets/16f0a677-29fd-45e6-b4a6-f0ca86e411c3" />

3. If you select a portion of the text that includes a box and delete, the box doesn't go back into the available authors section:

<img width="278" alt="Screenshot 2025-02-13 at 1 33 24 PM" src="https://github.com/user-attachments/assets/c543141a-ce2c-4e71-8d9f-ff33f7a52b0f" />

<img width="273" alt="Screenshot 2025-02-13 at 1 33 34 PM" src="https://github.com/user-attachments/assets/ad867b58-5fe0-41da-a125-363aabc5a866" />

4. Authors are always inserted at the end of the text box. I think the best experience would be for them to be inserted at the cursor position, otherwise it's tricky to edit existing bylines to add/change authors around.

leogermani commented:
This is starting to look pretty good!

Here are two things I noticed:

* This feature should also work without CoAuthors Plus. In that case, it would list only the normal post author
* The input should be pre-filled with the default by byline
* The value stored in the post meta field should follow what's described [here](https://github.com/Automattic/newspack-plugin/tree/trunk/includes/bylines#readme) 
 
The idea of storing the ID and the name is to have a fallback in case the user is deleted. When we render the byline, we'll always read the ID and get the Display name from the database, but if the ID is gone, we'll use the name that is stored there.

leogermani commented:
Ah, and before you get crazy and go deep in a rabbit role, we DON'T NEED TO SUPPORT Guest Authors

thomasguillot commented:
> Overall I think this is coming together! The UX of it feels good. I noticed a few quirks in my testing:
> 
> 1. In general it's kind of hard to read once there are a few authors in there:
> 
> <img alt="Screenshot 2025-02-13 at 1 29 49 PM" width="303" src="https://private-user-images.githubusercontent.com/7317227/413072138-d2d8b522-4afa-4ac2-9e85-401436d4595b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk1NDE0OTcsIm5iZiI6MTczOTU0MTE5NywicGF0aCI6Ii83MzE3MjI3LzQxMzA3MjEzOC1kMmQ4YjUyMi00YWZhLTRhYzItOWU4NS00MDE0MzZkNDU5NWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTRUMTM1MzE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTU4MjIzMGFjNThhM2IzOGY0ODk4MDc1MGY1MjRmN2E0YjJkZmY1YjQ0ZGY3NDVlNTQ2N2M4ZTU0MTcwOTM1YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.tyH__awef64mm6AlsL7slxML_MmCB3_52U4-9h3Qvsg">
> Not sure if we should reduce the size of the author boxes or how best to handle that. What do you think, @thomasguillot?

You're right, this doesn't look good. Let me think about a different approach here.

thomasguillot commented:
Note: The panel should simply be named "Byline", not "Newspack Custom Byline"

thomasguillot commented:
Can we please make sure the style matches Core's [FormTokenField](https://wordpress.github.io/gutenberg/?path=/docs/components-formtokenfield--docs)

Also we can use `@wordpress/base-styles/colors` rather than hardcoding the values.

allysonsouza committed:
fix: stylize tokens according to FormTokenField

allysonsouza committed:
fix: remove token when characterData changes

miguelpeixe commented:
I've just tested on a fresh post and I still have issues with the backspace interaction:

https://github.com/user-attachments/assets/9fa6593f-ded0-4c43-9de6-4f8b7e9eb816

Another problem I faced while drafting the new post was that the additional authors only became available after I saved and refreshed the post. Ideally, the component would be aware of the edited changes and update the component's available tokens.

allysonsouza committed:
fix: improved mutation observer callback logic to update data correctly

allysonsouza committed:
fix: initilize only on DOM ready and main byline element ready

allysonsouza committed:
fix: properly add the mutation observer after child component is ready

allysonsouza commented:
The latest commit fixes the backspace interaction, @miguelpeixe. About authors, it was expected that for the first version that would be acceptable (to have to refresh the page to make them available to use on byline), can you confirm that, @claudiulodro?

miguelpeixe commented:
Thanks for the revisions @allysonsouza!

Backspace interaction is working well for deleting the token when the cursor is at the end of it. I'm facing a couple of issues still:

 - The cursor repositions to the beginning after deleting a token
 - Removing (via delete or backspace) multiple selected text doesn't restore the token

https://github.com/user-attachments/assets/2422113b-59d1-4215-b371-46a468454e72

allysonsouza committed:
fix: implement setTokensInUse function to better reuse code

allysonsouza committed:
fix: split byline "textarea" and tokens to avoid unwanted re-rendering

allysonsouza committed:
feat: parse meta with <Author> before rendering byline

allysonsouza committed:
feat: store byline meta according to the expected format

allysonsouza committed:
chore: refacot bylineParser function

allysonsouza committed:
feat: make custom bylines to work without co-author plus activated

allysonsouza committed:
feat: make custom byline field to be displayed on a modal

allysonsouza committed:
fix: mystype newspackByline.is_co_authors_plus_active

allysonsouza committed:
fix: adding space after token insertion at the end of byline

allysonsouza committed:
fix: improve byline innerHTML assignment

leogermani commented on src/bylines/index.js:
```diff
@@ -3,64 +3,520 @@
+			</button>
+		</span>`;
+
+		// For backwards compatibility, replace the '<Author>' tags, which were used before.
```
no need for BC. This is brand new

leogermani committed:
docs: update docs

goldenapples commented on src/bylines/index.js:
```diff
@@ -3,64 +3,520 @@
+		<>
+			<Button variant="secondary" onClick={ openModal }>
+				Set Custom Byline
+			</Button>
```
It seems like there should be a way of seeing the byline without clicking to open the modal. Does it make sense to show the byline here rather than just the button? Is there a design for this?

goldenapples commented on src/bylines/index.js:
```diff
@@ -3,64 +3,520 @@
+	return (
+		<>
+			<span
+				className="components-form-token-field__token token-inline-block"
```
This probably needs `contenteditable="false"` since we're not trying to allow editing text within this span.

allysonsouza replied:
Nice suggestion, I didn't thought about that

goldenapples commented on src/bylines/index.js:
```diff
@@ -3,64 +3,520 @@
+	 * @return {string} Parsed byline looking up for <Author id=1></Author> tags and replacing them.
+	 */
+	const bylineParser = () => {
+		const tokenMarkup = `<span id="token-$1" class="components-form-token-field__token token-inline-block author-token" data-token="$1">
```
This looks like the same component as `TokenInlineBlock` - can they be refactored to use the same function for generating markup?

allysonsouza replied:
Yes, it is, but I couldn't think about how to retrieve the render of a component to store on a variable, I guess that's not possible, but another solution could be possible, I'm trying to think what.

goldenapples commented on src/bylines/index.js:
```diff
@@ -3,64 +3,520 @@
+		const query = { ...AUTHORS_QUERY };
+
+		return {
+			authors: getUsers( query ),
```
This is being fetched but not returned; is it necessary?

allysonsouza replied:
You're right, it was a reminiscence of a tentative or something, removed it

goldenapples commented on src/bylines/index.js:
```diff
@@ -3,64 +3,520 @@
+						.remove();
+
+					// Update byline meta.
+					updateBylineMeta( bylineElement );
```
Does this need to be called directly? I think removing the token from the field should trigger the mutation observer.

allysonsouza replied:
You're right, they don't seen to be necessary there, the mutation observer will be already triggered and it will call it

goldenapples commented on src/bylines/index.js:
```diff
@@ -3,64 +3,520 @@
+	 */
+	useEffect( () => {
+		// Wait for child component that will be analyzed to be ready.
+		if ( ! isTokensdReady || ! isBylineReady ) {
```
is this a typo "isTokensdReady"

goldenapples commented on src/bylines/style.scss:
```diff
@@ -13,3 +13,47 @@
+	height: auto;
+	line-height: 24px;
+	min-width: unset;
+	transition: all 0.2 scubic-bezier (0.4, 1, 0.4, 1);
```
Potential typo here in the easing function?

goldenapples commented on src/bylines/index.js:
```diff
@@ -3,64 +3,520 @@
+			const controller = new AbortController();
+
+			apiFetch( {
+				path: `/coauthors/v1/coauthors?post_id=${ postId }`,
```
Why can't we just use the edited coauthors attribute on the post?

allysonsouza committed:
fix: typo isTokensdReady

allysonsouza committed:
fix: remove non used authors query

allysonsouza committed:
fix: cubic-bezier transition typo

allysonsouza committed:
fix: remove updateBylineMeta from click on x/removal callback event

allysonsouza committed:
fix: remove unnecessary "backwards" compatibility since it's new feature

goldenapples committed:
refactor: replace MutationObserver with event listeners

Replaced the MutationObserver with an event listener, replaced some of
the DOM interaction with a useRef element, and tried to refactor some of
the data getters and setters.

goldenapples committed:
fix: update effect registration so it only runs on mounting

goldenapples committed:
refactor: remove unneded "remove token" handler

goldenapples committed:
fix: look up the byline meta on mount, not just on initial render

allysonsouza commented:
After struggling to handle the state of the content, which was not being managed by React state, @goldenapples started working on a simplification to use `onInput` instead of MutationObserver, which lead to a simpler and shorter code.

He's proposing it on [this PR on our fork](https://github.com/humanmade/newspack-plugin/pull/1), I've been testing it and it seems like we need to deal with the removal of the tokens itself when there's a change to the token content (like adding characters or removing them with backspace or delete).

That seems possible to be handled by onInput too, by adding an onInput on each token to remove itself when changes happen and to keep the onInput on the wrapper contenteditable div.

goldenapples committed:
fix: set contenteditable="false" on token spans

Prevents editing the content in these spans, and makes it so that
deleting (with backspace) will remove the whole span at once rather than
editing within the span.

goldenapples committed:
chore: add some missing docblocks and fix some lint issues

goldenapples committed:
fix: set default insert location to end of editable field

Updates the "cursorPos" variable logic to use the end of the line if
there's no cursor position set yet, and to add a non-breaking space at
the end of the text if necessary. Also updates the white-space styling
of the editable area, to ensure that users can select any point in the
text they may need to.

allysonsouza committed:
fix: remove margin from last token to prevent confusion with whitespace

goldenapples committed:
feat: add preview to sidebar interface

allysonsouza committed:
fix: rename panel to Byline

allysonsouza committed:
fix: make modal width fixed

allysonsouza committed:
feat: add function to parse meta for preview

leogermani commented:
Heads up that I've implemented the data structure changes that this PR does in trunk to allow us to move on with some tests. This might create conflicts that should be easily resolved - just ignore the changes on trunk

https://github.com/Automattic/newspack-plugin/pull/3863

allysonsouza committed:
feat: default custom byline on first toggle

joeleenk committed:
feat: set default custom byline meta on toggle

joeleenk committed:
feat: update styles to match new designs

joeleenk committed:
feat: update styles to match designs

joeleenk committed:
chore: cleanup, comments, organization

miguelpeixe reviewed (approved):
Thank you for all the revisions, the code is well organized and very readable!

I ran into a couple of issues, but I'm not considering them blockers because we can address them separately:

1. When backspacing a token, my cursor gets reset to position zero
2. Making changes to the post authors does not update the available tokens. This is also the case after saving the changes. Only after a page refresh I'm able to use the new authors. This only happens when using Co-Authors Plus.

Other than that, I'm able to make changes to the byline and save with different token/text variations without issues.

leogermani committed:
Merge branch 'trunk' into feat/byline-interface
